x/mint/keeper/keeper.go

```diff
@@ -101,3 +103,85 @@ func (k Keeper) MintCoins(ctx context.Context, newCoins sdk.Coins) error {
 func (k Keeper) AddCollectedFees(ctx context.Context, fees sdk.Coins) error {
 	return k.bankKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, k.feeCollectorName, fees)
 }
+
+func (k Keeper) DefaultMintFn(types.InflationCalculationFn) types.MintFn {
+	return func(ctx context.Context, env appmodule.Environment, minter *types.Minter, epochId string, epochNumber int64) error {
+		// the default mint function is called every block, so we only check if epochId is "block" which is
+		// a special value to indicate that this is not an epoch minting, but a regular block minting.
+		if epochId != "block" {
+			return nil
+		}
+
+		stakingTokenSupply, err := k.StakingTokenSupply(ctx)
+		if err != nil {
+			return err
+		}
+
+		bondedRatio, err := k.BondedRatio(ctx)
+		if err != nil {
+			return err
+		}
+
+		params, err := k.Params.Get(ctx)
+		if err != nil {
+			return err
+		}
+
+		minter.Inflation = minter.NextInflationRate(params, bondedRatio)
+		minter.AnnualProvisions = minter.NextAnnualProvisions(params, stakingTokenSupply)
+
+		mintedCoin := minter.BlockProvision(params)
+		mintedCoins := sdk.NewCoins(mintedCoin)
+		maxSupply := params.MaxSupply
+		totalSupply := stakingTokenSupply
+
+		// if maxSupply is not infinite, check against max_supply parameter
+		if !maxSupply.IsZero() {
+			if totalSupply.Add(mintedCoins.AmountOf(params.MintDenom)).GT(maxSupply) {
+				// calculate the difference between maxSupply and totalSupply
+				diff := maxSupply.Sub(totalSupply)
+				if diff.LTE(math.ZeroInt()) {
+					k.logger.Warn("max supply reached, no new tokens will be minted")
```

Comment:
personal preference: info level would be fine, too. There are no actionable steps for an operator. The warning level can be annoying.

Reply:
I'll apply this suggestion, thank you